### ISSUE TITLE
[Discussion] Alert usage in IPs

### ISSUE THREAD

**Issue by weicaiyang** [lowRISC/opentitan#3767]
As mentioned in #3742, would like to further discuss all kinds of alert usage in IPs

I'm thinking if we can define all kinds of alert behavior in comportability spec. I try to understand the alert behavior in current IPs.
In OTP, alert has err_code, but it's RO. Looks like it's one time triggered. LMK if it's wrong.
In keymgr, alert is constantly triggered until SW clears err_code and err_code is W1C
In AES, there are 2 alerts: ctrl_err_update and ctrl_err_storage. Each alert has one kind of error, so there is no err_code. ctrl_err_update is one time triggered, while ctrl_err_storage is constantly triggered until reset (SW can't clear the alert, unless it's just a glitch). Not sure if it's better to make it constantly triggered even it's a glitch. @vogelpi

Hence, we have 3 kinds of alert now

1. one time triggered. SW doesn't need to clear it. Just need alert handler ACK it and report to system
2. constantly triggered until SW clears err_code. error_code is one bit one error in keymgr, but it's not in OTP, what happens if more than 1 error occur?
3. constantly triggered until reset

For interrupt, we have interrupt pin, intr_state, intr_enable and intr_test. Do we need to have alert_test as well (probably don't need alert_enable as it should not be disabled)?

**Comment by msfschaffner** [comment]
Thanks for bringing this up, @weicaiyang! This is an important topic.

Note that as long as the `alert_i` input of an alert sender remains high, the alert sender will repeatedly send out alert events (the re-send period is dependent on the bidirectional handshake duration and hence also the clock ratios of the sender and receiver domains). So if not used correctly, we may get a large number of alert events on the alert handler side even in cases where the actual alert condition was only true for one cycle. This may distort alert accounting in the classifier and alert accumulator.

Hence, we could require that the `alert_i` input of the alert sender should only be pulsed high as long as an alert condition is true - which in many cases is just one cycle. Only in cases where the alert condition continues to persist should `alert_i` remain asserted. We can also rename the `alert_i` to `alert_event_i` to make this more clear.

That would mean that most alerts should probably fall into category 1, except for cases where an alert condition continuously persists (which in my view should rather fall into category 3). Hence, while it is ok to derive alerts from error conditions, I think in most cases we probably do not want the `alert_i` signal to remain asserted until the error CSRs are cleared by SW...

------

As for multiple alert events: Note that the alert handler has not been designed to count the alert events exactly. Alerts events may get squashed in the sender/receiver, or in the alert classification if they occur to close to one another.

-----

RE the test register: Indeed, this is something that we have tasked for alert_handler D2. This will likely just be a trigger register that feeds a pulse into `alert_i`, if written to by SW.

Other thoughts?

**Comment by tjaychen** [comment]
so i actually almost filed an issue on this yesterday, but decided I'd be
okay letting alerts trigger until software responds.
But Michael, if we want to send only one, shouldn't we technically keep the
input alert_i asserted until the alert sender knows for sure it sent the
alert and it was received?

I think my main worry is if it's a singular event, it may be attackable.
What do you think?
And let me know if I misunderstood how it's constructed.

**Comment by tjaychen** [comment]
i should explain my motivation a bit.
I have an error register with unique bits. And basically wanted each to be
able to trigger an alert.
I don't mind managing this outside, but I realized I don't have a good idea
when an alert has actually "gone" out, and therefore I need to keep
registering what has already been sent and what has not.

I dont think this particular case warrants separate alerts.

**Comment by weicaiyang** [comment]
> As for multiple alert events: Note that the alert handler has not been designed to count the alert events exactly. Alerts events may get squashed in the sender/receiver, or in the alert classification if they occur to close to one another.

@msfschaffner 
what if multiple errors happen in OTP? The ERR_CODE_x can only show one error at a time. keymgr has unique bit for each error, so SW can know when there are multiple errors.

**Comment by vogelpi** [comment]
Thanks for pinging me @weicaiyang . In aes, we have indeed just two alerts for the two error conditions you mentioned. They map to cases (1) and (3).

As @msfschaffner explained, the alert sender makes sure every alert is signaled to and acknowledged by the receiver, even if it's just a pulse (Case 1). So there is no need to keep it asserted if we just want to signal one alert. Regarding attacking the singular event: If we register the alert condition to keep the alert input asserted, the adversary can "just" attack the input to the flop (which may still be a singular event). I don't think we would win much here. If we really care, we must make sure the alert condition is continuously generated and we can't get out of it. This is what we do for the `ctrl_err_storage` in aes (Case 3).

For aes, we might add more error conditions and group them into a minor (recoverable) and major (non-recoverable) alert in the future. This is what we do in Ibex (not yet connected to the alert handler though). I think this might even be required at some point to limit the number of alerts and avoid scalability issues (see interrupt consolidation/PLIC discussions). Then, we should probably put an edge detector in front of the alert input that generates one pulse for every non-to-zero change in the error code. Some errors in the error code can still go to the major alert and be set continuously. In addition, the error code register should probably have one bit per error as in keymgr to let software know what happened.

**Comment by weicaiyang** [comment]
> Thanks for pinging me @weicaiyang . In aes, we have indeed just two alerts for the two error conditions you mentioned. They map to cases (1) and (3).
> 
> As @msfschaffner explained, the alert sender makes sure every alert is signaled to and acknowledged by the receiver, even if it's just a pulse (Case 1). So there is no need to keep it asserted if we just want to signal one alert. Regarding attacking the singular event: If we register the alert condition to keep the alert input asserted, the adversary can "just" attack the input to the flop (which may still be a singular event). I don't think we would win much here. If we really care, we must make sure the alert condition is continuously generated and we can't get out of it. This is what we do for the `ctrl_err_storage` in aes (Case 3).
> 
> For aes, we might add more error conditions and group them into a minor (recoverable) and major (non-recoverable) alert in the future. This is what we do in Ibex (not yet connected to the alert handler though). I think this might even be required at some point to limit the number of alerts and avoid scalability issues (see interrupt consolidation/PLIC discussions). Then, we should probably put an edge detector in front of the alert input that generates one pulse for every non-to-zero change in the error code. Some errors in the error code can still go to the major alert and be set continuously. In addition, the error code register should probably have one bit per error as in keymgr to let software know what happened.

Thanks for the comment. AES 2 alerts make sense to me.

**Comment by tjaychen** [comment]
> Thanks for pinging me @weicaiyang . In aes, we have indeed just two alerts for the two error conditions you mentioned. They map to cases (1) and (3).
> 
> As @msfschaffner explained, the alert sender makes sure every alert is signaled to and acknowledged by the receiver, even if it's just a pulse (Case 1). So there is no need to keep it asserted if we just want to signal one alert. Regarding attacking the singular event: If we register the alert condition to keep the alert input asserted, the adversary can "just" attack the input to the flop (which may still be a singular event). I don't think we would win much here. If we really care, we must make sure the alert condition is continuously generated and we can't get out of it. This is what we do for the `ctrl_err_storage` in aes (Case 3).
> 
> For aes, we might add more error conditions and group them into a minor (recoverable) and major (non-recoverable) alert in the future. This is what we do in Ibex (not yet connected to the alert handler though). I think this might even be required at some point to limit the number of alerts and avoid scalability issues (see interrupt consolidation/PLIC discussions). Then, we should probably put an edge detector in front of the alert input that generates one pulse for every non-to-zero change in the error code. Some errors in the error code can still go to the major alert and be set continuously. In addition, the error code register should probably have one bit per error as in keymgr to let software know what happened.

Yes while this is all true, it would still be nice to have a way to sequence the input alert events.  For example, if I have multiple error conditions generate, I might still prefer to send N number of alerts instead of OR-ing them into one.  That to me is why it would be helpful to have an indication from the alert sender that the last "alert" has been processed. If the IP doesn't care that alerts are OR'd together, it doesn't have to make use of that information.  But in case it does care, at least there will be an option.

**Comment by msfschaffner** [comment]
Ok I see. The alert sender has a set-register at the input to latch a pulse, and this set-register will be cleared once the alert went out. We could output this information and make the interface behave like a req/ack interface so that alert events become sequence-able (IPs that don't care could then just ignore the ack signal). The change should not be too invasive, assuming that the DV testbench will ignore the ack output (we should be able to cover this output in FPV). 

Note however that the alert handler itself uses an OR reduction internally to aggregate alerts within a class for simplicity. So there is still a chance that multiple alert events from different sources get combined into one event.

CC @cindychip 
------------------------------------------------------

@weicaiyang: the OTP errors are a bit special. there exist many error codes, but most of them cannot appear simultaneously (e.g. write errors and ECC read errors are mutually exclusive). there are just a few that can occur simultaneously with others (e.g. a concurrent parity error or escalation), and these have precedence when it comes to error code assignment.

Note also that there are multiple agents inside the OTP (e.g. a partition, or the programming interface) that operate in parallel. That is why each of those agents has its own error status bit and error code register.

The current OTP implementation only generates single alert events at the moment. But I am thinking about changing that such that the terminal error conditions keep on sending alerts until reset.

**Comment by cindychip** [comment]
Thanks for tagging me @msfschaffner @tjaychen, I think 
1. The set-register in alert_sender could be tested in FPV as michael mentioned.
2. For req_ack signals between alert_sender and IP (who triggered the alert), this will be tested in individual IPs if needed, not alert_handler TB itself.
So in short, both proposals would not affect alert_handler tb.

**Comment by weicaiyang** [comment]
> @weicaiyang: the OTP errors are a bit special. there exist many error codes, but most of them cannot appear simultaneously (e.g. write errors and ECC read errors are mutually exclusive). there are just a few that can occur simultaneously with others (e.g. a concurrent parity error or escalation), and these have precedence when it comes to error code assignment.
> 
> Note also that there are multiple agents inside the OTP (e.g. a partition, or the programming interface) that operate in parallel. That is why each of those agents has its own error status bit and error code register.
> 
> The current OTP implementation only generates single alert events at the moment. But I am thinking about changing that such that the terminal error conditions keep on sending alerts until reset.

sounds good. Thanks for the explanation.

**Comment by msfschaffner** [comment]
I've made a prototype for automatically generated alert test registers in #3922. PTAL.

**Comment by weicaiyang** [comment]
@msfschaffner @tjaychen 

can we conclude this issue?
alert test has been added in #3922.
In #3817, we changed `alert_i` to `alert_req_i`, which allows alert req to last until alert sender acknowledges it, but it's also OK to drive `alert_req_i` as a pulse and ignore the `alert_ack_o` from alert sender.

And looks like we don't need the alert usage case2. 
1. one time triggered. Here are 2 situations.
  a. the alert is only for one error. SW doesn't need to clear it. Just need alert handler ACK it and report to system
  b. multiple errors may cause the alert. The error is shown in CSR err_code. Now err_code is W1C in keymgr. SW should clear it otherwise, when err_code is changed to the other non-zero value, this alert will occur again. But in OTP, this is RO. Looks like SW doesn't need to clear it. Do we need to unify the reg type for err_code?
2. ~constantly triggered until SW clears err_code. error_code is one bit one error in keymgr, but it's not in OTP, what happens if more than 1 error occur?~
3. constantly triggered until reset

**Comment by msfschaffner** [comment]
Thanks for keeping track of this @weicaiyang. 

I believe the only open question remaining is whether we should standardize the error / alert register type (your question in 1b). I feel like we should leave some flexibility there since, since some IPs may deal with errors / alerts in different ways. 

As for OTP, I should add that all alerts within OTP are terminal, which means that they fall into category 3) above. This is the reason why the associated error regs are RO.

So IMO I think we could standardize on the following guidance.

---------------------------------------------------
Alerts of comportable IPs in the system must be in either of the following two categories:

1) One-time triggered alerts. This category is for regular alerts that are due to recoverable error conditions. The alert sender shall send out one single alert event when the corresponding error condition is asserted.
2) Alerts that are continuously triggered until reset. This category is for highly critical alerts that are due to terminal error conditions. The alert sender shall continuously send out alert events as long as the corresponding error condition is true.

If the alert is only for one error source, the IP may expose an associated error register - but this is not a requirement. If there are multiple error sources per alert, the IP shall expose an associated error register to indicate the cause.

Error registers for one-time triggered alerts (1) must either be clearable by SW, or the HW must provide an automatic mechanism to clear them (e.g., upon starting a new transaction initiated by SW).

Error registers for continuously triggered alerts (2) must not be clearable in any way and must hence be read-only.

-------------------------------------
@tjaychen what is your take?

Btwy, do we want to document this in a section on alerts in our [comportability definition/spec doc](https://docs.opentitan.org/doc/rm/comportability_specification/)?

**Comment by weicaiyang** [comment]
@msfschaffner This guidance looks great to me.

 > Error registers for one-time triggered alerts (1) must either be clearable by SW, or the HW must provide an automatic mechanism to clear them (e.g., upon starting a new transaction initiated by SW).

If there is no error register for the one-time triggered alert, it's also fine that we don't need SW to clear anything, right? Like `ctrl_err_update` in AES, this alert is in category (1) and it doesn't have corresponding error register since it associates with one alert. When it happens, no need to do anything in AES module.

**Comment by msfschaffner** [comment]
Yes the guidance is meant to imply that this is only needed in case there is an error register. If there is no error register, SW does not have to take any action.

**Comment by weicaiyang** [comment]
> Yes the guidance is meant to imply that this is only needed in case there is an error register. If there is no error register, SW does not have to take any action.

Thanks for clarification. @msfschaffner

**Comment by tjaychen** [comment]
yes i think we should standardize this, and in general i agree with @msfschaffner outline.
We should probably run this through the silicon meeting to see if everyone generally agrees.

**Comment by vogelpi** [comment]
Thanks @msfschaffner for the proposal. This sounds good to me.

**Comment by vogelpi** [comment]
@msfschaffner gave a nice presentation on this topic in the last silicon meeting ([slides on the OpenTitan team drive](https://docs.google.com/presentation/d/1hr-vSEgqOju2EeWDEFN_02zHuwLok79IFuzAC97bLZw/edit?resourcekey=0-fuK707X2J6FKswZqsHOngg#slide=id.p)).

During the Q&A session as well as in PR https://github.com/lowRISC/opentitan/pull/4781#discussion_r554118992 we have been discussing that we should probably guard the error/cause register giving more details on the alert source with a lifecycle signal. DEV, RMA, TEST_UNLOCKED were mentioned as states in which those regs should be accessible. It was also mentioned that we should get more input on this from the labs.

I am adding this information here because we should take a decision together, document the guideline and adhere to it consistently. If preferred, we can also fan out the error/cause reg guarding to a separate issue.

**Comment by msfschaffner** [comment]
Thanks for adding this item, @vogelpi. I split it out into a separate issue #4902 such that we can close this one.

The formalized alert guidance is coming in #4891 and should get merged soon.

**Comment by msfschaffner** [comment]
Ok the new guidance is in, and the alert sender primitive has been updated accordingly in #4863.
Block owners of the following IPs, please align the naming and usage, and tick the box below once done:

- [ ] aes
- [ ] otbn
- [ ] keymgr
- [ ] sensor_ctrl
- [ ] sram_ctrl (#4864)
- [ ] lc_ctrl (#4864)
- [ ] otp_ctrl (#4864)

CC @tjaychen @vogelpi @imphil @rswarbrick @msfschaffner

**Comment by tjaychen** [comment]
do you think there's any way to add a check into topgen to see whether all
alerts are following the naming convention?
once these are done, we can make that an error directly in the future.

**Comment by msfschaffner** [comment]
good idea, this should be easy to add.
shall we just make it a warning for the time being until everything is aligned then?

**Comment by tjaychen** [comment]
yeah i think that would be a good idea.

**Comment by mwbranstad** [comment]
- [x ] csrng
- [x ] edn
- [x ] etnropy_src

All alerts added (#5195,#5234,#5284).

**Comment by weicaiyang** [comment]
Looks like alert name of keymgr and sensor_ctrl have been aligned.
Perhaps we can close this issue now. @msfschaffner WDYT?

> Ok the new guidance is in, and the alert sender primitive has been updated accordingly in #4863.
> Block owners of the following IPs, please align the naming and usage, and tick the box below once done:
> 
> * [x]  aes (#4909)
> * [x]  otbn
> * [x]  entropy_src
> * [ ]  keymgr
> * [ ]  sensor_ctrl
> * [x]  sram_ctrl (#4864)
> * [x]  lc_ctrl (#4864)
> * [x]  otp_ctrl (#4864)
> 
> CC @tjaychen @vogelpi @imphil @rswarbrick @msfschaffner @mwbranstad

**Comment by msfschaffner** [comment]
Yes this LGTM. Let's close this issue down.
